File: 79-word-search/word-search.cpp
class Solution {
public:

    bool dfs(vector<vector<char>> & board, string& word, int i, int j, int idx, vector<vector<bool>>& vis){
        if (idx == word.size()) return true;
        int m = board.size(), n = board[0].size();
        if(i<0 || i>=m || j<0 || j>=n || board[i][j]!=word[idx] || vis[i][j]) return false;

        vis[i][j] = true;
        bool found = 
        dfs(board , word, i+1, j, idx+1, vis) ||
        dfs(board , word, i, j+1, idx+1, vis) ||
        dfs(board , word, i-1, j, idx+1, vis) ||
        dfs(board , word, i, j-1, idx+1, vis);
        vis[i][j] = false;
        return found;
    }
   
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size(), n = board[0].size();
        vector<vector<bool>> vis(m, vector<bool>(n, false));

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == word[0] && dfs(board, word, i, j, 0, vis))
                    return true;
            }
        }
        return false;
    }
};

File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        row , col = len(board), len(board[0])
        visited = [[False for _ in range(col)] for _ in range(row)]
        def dfs(idx, r, c):
            if idx == len(word): return True
            if r>=row or c>=col or r<0 or c<0 or visited[r][c] or word[idx]!=board[r][c]: return False
            visited[r][c] = True
            possible = (
                dfs(idx+1, r+1 , c) or 
                dfs(idx+1, r-1, c) or 
                dfs(idx+1, r, c-1) or dfs(idx+1, r, c+1)
                )
            visited[r][c] = False
            return possible
        for r in range(row):
            for c in range(col):
                if word[0] == board[r][c] and dfs(0, r, c):
                    return True
        return False